File: solutions/python/bank-account/1/bank_account.py



class BankAccount:
    def __init__(self):
        self.balance = 0
        self.status = "closed"

    def get_balance(self):
        if self.status == "closed":
            raise ValueError("account not open")
        return self.balance

    def open(self):
        if self.status == "opened":
            raise ValueError("account already open")
        self.status = "opened"

    def deposit(self, amount):
        if self.status == "closed":
            raise ValueError("account not open")
        if amount < 0:
            raise ValueError("amount must be greater than 0")
        self.balance += amount

    def withdraw(self, amount):
        if self.status == "closed":
            raise ValueError("account not open")
        if amount > self.balance:
            raise ValueError("amount must be less than balance")
        if amount < 0:
            raise ValueError("amount must be greater than 0")
        self.balance -= amount

    def close(self):
        if self.status == "closed":
            raise ValueError("account not open")
        self.status = "closed"
        self.balance = 0

File: solutions/java/bank-account/1/src/main/java/BankAccount.java
class BankAccount {

    private boolean isOpen;
    private int balance;

    void open() throws BankAccountActionInvalidException {
        if (this.isOpen)
            throw new BankAccountActionInvalidException("Account already open");
        this.isOpen = true;
        this.balance = 0;
    }

    void close() throws BankAccountActionInvalidException {
        if (!this.isOpen)
            throw new BankAccountActionInvalidException("Account not open");
        this.isOpen = false;
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        if (!this.isOpen)
            throw new BankAccountActionInvalidException("Account closed");
        return this.balance;
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        if (!this.isOpen)
            throw new BankAccountActionInvalidException("Account closed");
        if (amount < 0)
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
        this.balance += amount;
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        if (!this.isOpen)
            throw new BankAccountActionInvalidException("Account closed");
        if (amount < 0)
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
        if (amount > this.balance)
            throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
        this.balance -= amount;
    }

}
